PyTorch-Sparse-Roadmap.md

```diff
@@ -0,0 +1,515 @@
+`torch.matmul`, `torch.sparse.mm`, `torch.sparse.FloatTensor.mm`, etc.).
+There should be a thorough pass through the existing PyTorch codebase to ensure
+that any code relying on sparse tensors still functions as intended with the
+changes made. A suitably robust suite of unit tests should be designed to verify
+that correct behavior is preserved with any modifications to the code. This
+would include ensuring GPU support as well as compliance with `autograd`.
+
+## Related Work
+
+Much prior work on sparse matrices is implemented in the 
+[`scipy.sparse`](https://docs.scipy.org/doc/scipy/reference/sparse.html) 
+module; the [`pydata/sparse`](https://github.com/pydata/sparse) project extends 
+some of those ideas to sparse arrays of generic rank `n`. As a stretch goal, it 
+may make sense to have general rank `n` PyTorch tensors being supported in most 
+of the same ways as `pydata/sparse` and to have sparse tensors of rank 2
+supported in most of the same ways as `scipy.sparse`. 
+
+Among projects tailored to neural networks & deep learning,
+[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul)
+has a `SparseTensor` class that uses COO format for tensors of arbitrary
+rank `n` and a `SparseMatrix` class that uses CSR format for tensors of
+rank 2; explicit conversion routines between formats exist where they apply. 
+There are also rank 2 `CSRNDArray` and rank 1 `RowSparseMatrix` classes 
+included in
+[MXNet](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseMatrixSparseMatMul).
+The project [`pytorch_sparse`](https://github.com/rusty1s/pytorch_sparse)
+suggests some directions for supporting efficient sparse matrix operations
+with `autograd` and also with CUDA support.
+
+Finally, there are generalizations of sparse matrix storage formats to sparse 
+tensors of arbitrary rank `n` described in various articles listed in the 
+references.
+
+## Implementation
+
+The module `torch` already contains an attribute `torch.sparse_coo` of type
+`torch.layout` with an associated constructor `torch.sparse_coo_tensor`. The
+proposed CSR implementation should extend the API by adding a few attributes
+to the `torch` module with similar names:
+
++ `torch.sparse_csr` (an attribute of type `torch.layout`);
++ `torch.sparse_csr_tensor` (an associated constructor); and
++ functions `torch.to_sparse_csr` & `torch.to_sparse_coo` for explicit
+  conversion between representations (for rank 2 sparse tensors).
+
+The `torch.sparse_csr_tensor` constructor would have a similar signature to `torch.sparse_coo_tensor`, i.e.,
+
+```{python}
+sparse_csr_tensor(indices, values, size=None, dtype=None,
+                  device=None, requires_grad=False)       -> Tensor
+```
+
+where `indices` would be a list of two lists (e.g., `row_index` & `col_index`
+from the example before) and `values` would be a tensor containing the
+corresponding nonzero entries of the sparse tensor.
```

Comment:
Remove "nonzero" or replace it with "presumably nonzero" as nothing would check if the entries will be nonzero (it just would be too expensive to do that).

Should we be explicit here on whether the dense tensor of `values` can be an N-dimensional tensor o(similar to the COO case) or not?

Reply:
Fixed.